packages/blade/src/components/Input/TextInput/useFormattedInput.ts

```diff
@@ -0,0 +1,186 @@
+      if (didDelete) {
+        const deletedChar = currentValue[cursorPosition] ?? ''; // "12/34"[2] → "/"
+        const deletedDelimiter = !isUserCharacter(deletedChar); // "/" → true (is delimiter)
+
+        if (deletedDelimiter) {
+          // true (will execute for "/" deletion)
+          const beforeCursor = newInputValue.substring(0, cursorPosition); // "12" (before cursor)
+          const afterCursor = newInputValue.substring(cursorPosition); // "34" (after cursor)
+          const rawBefore = stripPatternCharacters(beforeCursor); // "12" → "12"
+          const rawAfter = stripPatternCharacters(afterCursor); // "34" → "34"
+
+          rawValue = rawBefore.slice(0, -1) + rawAfter; // "12".slice(0,-1) + "34" → "1" + "34" → "134"
+
+          // Removes trailing non-alphanumeric characters from the end of the string, preserving the last alphanumeric word before them.
+          infoRef.current.cursorPosition =
+            beforeCursor.replace(/([\d\w]+)[^\dA-z]+$/, '$1').length - 1;
+        }
+      }
+
+      const formattedValue = format(rawValue, pattern); // format("134", "##/##") → "13/4"
+      infoRef.current.endOfSection = false;
+
+      // Handle cursor positioning when typing (not deleting)
+      if (!didDelete) {
+        // User types "2" in "1|" → becomes "12|/" → should jump to "12/|"
+        const nextChar = formattedValue[cursorPosition]; // "12/"[2] → "/" (delimiter)
+        const nextIsDelimiter = nextChar ? !isUserCharacter(nextChar) : false; // "/" → true
+
+        const remainingText = formattedValue.substring(cursorPosition); // "12/".substring(2) → "/"
+        const nextUserCharIndex = remainingText.search(/[\dA-z]/); // "/".search() → -1 (no user chars)
+        const hasMoreUserChars = nextUserCharIndex !== -1; // -1 !== -1 → false
+
+        infoRef.current.endOfSection = nextIsDelimiter && !hasMoreUserChars; // true && false → false
+
+        // Move cursor past auto-inserted delimiters for smooth typing
+        if (nextIsDelimiter && hasMoreUserChars) {
+          const prevChar = formattedValue[cursorPosition - 1] ?? '';
+          const prevIsDelimiter = !isUserCharacter(prevChar);
+
+          if (prevIsDelimiter) {
+            infoRef.current.cursorPosition = cursorPosition + nextUserCharIndex + 1;
+          }
+        }
+      }
+
+      onChange?.({ name, value: formattedValue, rawValue });
+      setInternalValue(formattedValue);
+    },
+    [pattern, onChange, internalValue],
+  );
+
+  // Apply calculated cursor position after value updates
+  useEffect(() => {
+    const { cursorPosition, endOfSection } = infoRef.current;
+
+    if (endOfSection || cursorPosition === undefined) return; // Skip if no position or end section
+
+    if (inputRef.current) {
+      inputRef.current.setSelectionRange(cursorPosition, cursorPosition);
+    }
+  }, [internalValue]);
+
+  const handleKeyDown = useCallback((event: React.KeyboardEvent<HTMLInputElement>) => {
```

Comment:
Can we restructure such that all the event handlers are at once place 1 by 1 and all the effects are in 1 place? 

useEffect
useEffect
handleSomething
handleKeydown